### PR TITLE
sql: exponential functions for apd

### PR DESCRIPTION
**Pull request by sploiselle**
More built-in APD functions

### CODE REVIEW

**Review by benesch — APPROVED** [review]
LGTM but didn't review for correctness.

**Comment by philip-stoev** [comment]
I had to bump "dec" to 0.4.2 as otherwise ``` SELECT exp('2'::apd)``` was crashing for me:

```2021-06-10T06:57:52.549262Z ERROR panic: range end index 14 out of range for slice of length 13
thread: coordinator
location: /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/dec-0.4.0/src/ordered.rs:222:9
```

File: src/expr/src/scalar/func.rs [hunk 1]
```diff
@@ -1457,9 +1518,34 @@ fn exp_dec<'a>(a: Datum<'a>, scale: u8) -> Result<Datum<'a>, EvalError> {
...
+        Err(EvalError::FloatOverflow)
+    } else if cx_status.subnormal() {
+        Err(EvalError::FloatUnderflow)
+    } else {
+        apd::munge_apd(&mut a.0).unwrap();
+        Ok(Datum::APD(a))
+    }
+}
+
 fn power<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
```
Comment (philip-stoev) [inline src/expr/src/scalar/func.rs:1536]: ```POWER (0,0)``` appears to be problematic (https://en.wikipedia.org/wiki/Zero_to_the_power_of_zero#Treatment_on_computers), but the consensus between Postgres, MySQL, SQL Server and Oracle is that the answer should be ```1``` . We return```NaN```:

```
materialize=> SELECT POW ( '0'::APD, '0'::apd);
 pow 
-----
 NaN
(1 row)
```

**Review by philip-stoev — APPROVED** [review]
Thank you, the randomized test shows that we return results identical to Postgres, except for the ```POW(0,0)``` case . We also return equivalent error messages for the impossible cases , e.g. logarithm of zero and so on.

File: src/expr/src/scalar/func.rs (hunk 1 above)
Comment (sploiselle) [reply src/expr/src/scalar/func.rs:1627]: Oh, great catch--tysm. Fixed this.

**Pull request merged**